substrate-node/pallets/pallet-smart-contract/src/lib.rs

```diff
@@ -842,52 +908,117 @@ impl<T: Config> Pallet<T> {
+                    block
+                );
             }
-            let mut contract =
-                Contracts::<T>::get(contract_id).ok_or(Error::<T>::ContractNotExists)?;
+            Err(err) => {
+                // if billing failed append it to the failed ids in local storage
+                // billing will be retried at the next block
+                log::error!(
+                    "error while billing contract with id {:?}: <{:?}>",
+                    contract_id,
+                    err.error
+                );
+                Self::append_failed_contract_ids_to_storage(contract_id);
+                return Ok(().into());
+            }
+        }
 
-            // Try to bill contract
-            match Self::bill_contract(&mut contract) {
-                Ok(_) => {
-                    log::info!(
-                        "billed contract with id {:?} at block {:?}",
-                        contract_id,
-                        block
-                    );
-                }
-                Err(err) => {
-                    log::info!(
-                        "error while billing contract with id {:?}: {:?}",
-                        contract_id,
-                        err
-                    );
-                    // If billing the contract failed, we should delete the contract and clean up storage
-                    Self::remove_contract(contract.contract_id);
-                    continue;
-                }
+        // https://github.com/threefoldtech/tfchain/issues/264
+        // if a contract is still in storage and actively getting billed whilst it is in state delete
+        // remove all associated storage and continue
+        let ctr = Contracts::<T>::get(contract_id);
+        if let Some(contract) = ctr {
+            if contract.contract_id != 0 && contract.is_state_delete() {
+                Self::remove_contract(contract.contract_id);
+                return Ok(().into());
             }
 
-            // https://github.com/threefoldtech/tfchain/issues/264
-            // if a contract is still in storage and actively getting billed whilst it is in state delete
-            // remove all associated storage and continue
-            let ctr = Contracts::<T>::get(contract_id);
-            if let Some(contract) = ctr {
-                if contract.contract_id != 0 && contract.is_state_delete() {
-                    Self::remove_contract(contract.contract_id);
-                    continue;
-                }
+            // Reinsert into the next billing frequency
+            Self::_reinsert_contract_to_bill(contract.contract_id);
+        }
+
+        Ok(().into())
+    }
+
+    fn bill_contract_using_signed_transaction(
+        block_number: T::BlockNumber,
+        contract_id: u64,
+    ) -> Result<(), Error<T>> {
+        let signer = Signer::<T, T::AuthorityId>::any_account();
+        let result = signer.send_signed_transaction(|_acct| Call::bill_contract_for_block {
```

Comment:
best to check if the signer can actually sign before making the call. 

I believe there is a method on signer: `signer.can_sign()`

Reply:
I think the send_signed_transaction will fail if you cannot sign no?

Reply:
I've seen other some other offchain worker code where they checked this before actually trying, might be good practice to do it as well